### DISCUSSION TITLE
Interesting 10-day schedule problem.

### DISCUSSION THREAD

**Discussion by gAlleb** [savonet/liquidsoap#4028] [General]
Hi! So

> We have twice daily sessions, one in the AM around 8am and another in the afternoon about 4pm. We play the same audio for both the AM and PM sessions on that day.
There are 10 audios in our program that are played in sequence over the course of 10 days. After the last session, we start all over again on another 10-day sequence. 

I've been thinking about:

```ruby

playlist_10_8am = playlist(mode="normal", loop=true, "/playlists/playlist_10_8am.m3u")

playlist_10_16pm = playlist(mode="normal", loop=true, "/playlists/playlist_10_16pm.m3u")

radio = mksafe(switch(
      [(predicate.once({ 8h00m }), playlist_10_8am), 
       (predicate.once({ 16h00m }), playlist_10_16pm) 
       ]
    ))
```

So now we have 2 playlists, each one contains 10 identical files, they are sequential (mode='normal'), they are not watched.
We have a `switch`, which plays everyday 1 file from 1st playlist at 8:00 and 1 file from 2nd playlist at 16:00 and silence in between. When 10 days are gone - it all starts over).

I see how it's not ideal at all) Cause we have to start LS specifically on a day when 10-day program starts. Some thinking should be made). Maybe something with `time()`.

How we can achieve expected behavior more elegantly?

**Comment by vitoyucepi** [comment]
Hi @gAlleb,
First of all, I'd like to discuss the possibility of implementing the required functions and scheduling in the external script. I think pushing to the queue using the API is much easier than storing the state using Liquidsoap's internal functions.

**Reply by gAlleb** [reply]
Yeah, totally cool.

Something with your altered `request.dynamic`
```ruby
def get_request()
  command = process.quote.command("------", args=[------])
  uri =
    try
      list.hd(process.read.lines(command))
    catch _ do
      null()
    end

  null.map(fun(uri) -> request.create(uri, temporary=true), uri)
end

def replaces request.dynamic(%argsof(request.dynamic), callback)
  prevent = ref(false)

  def disable_prevent()
    prevent := false
  end

  def enable_prevent()
    prevent := true
    delay = max(getter.get(retry_delay) - 0.1, 0.1)
    thread.run(disable_prevent, delay=delay)
  end

  calls = ref(0)
  max_calls = 3

  def reset_calls()
    calls := 0
  end

  def increment_calls()
    calls := calls() + 1
  end

  thread.run(reset_calls, every=retry_delay, fast=true)

  def debounce()
    if prevent() then
      null()
    elsif calls() >= max_calls then
      null()
    else
      result = callback()
      if result == null() then
        enable_prevent()
      end
      increment_calls()
      result
    end
  end

  request.dynamic(debounce, %argsof(request.dynamic))
end

requests = request.dynamic(get_request, id="requests", retry_delay=10.)
```

or via API. 

> than storing the state using Liquidsoap's internal functions.

I do agree on that. Except the fact I have never implemented something like that.

**Reply by gAlleb** [reply]
So basically we are taking about

```ruby
10_day_program = request.dynamic(get_request, id="10_day_program", retry_delay=10.)
radio = mksafe(switch(
      [(predicate.once({ 8h00m }), 10_day_program), 
       (predicate.once({ 16h00m }), 10_day_program) 
       ]
    ))
```

And creating a script which does the calculations what file to output when it's called depending on a date. But what to choose? PHP, python, JS? It becomes more complex due to the need to become more proficient in the required language))

**Reply by vitoyucepi** [reply]
I'm thinking about an HTTP API that will receive URLs of the tracks and push them to `request.queue`.
```ruby
q1 = request.queue()
radio = fallback([q1, radio])
```

> But what to choose? PHP, python, JS?

Use the language you are most/least familiar with.

**Reply by gAlleb** [reply]
> implementing the required functions and scheduling in the external script. I think pushing to the queue using the API ... that will receive URLs of the tracks and push them to request.queue.

I see. Could you please write me full logic (a bit detailed algorithm per say), not a script (code), but in your own general words so I then could try to implement it myself (e.g. what should happen in external script and in LS script).

**Reply by vitoyucepi** [reply]
Sure.
You should create a separate script that will handle time and date logic and also store the state of played tracks.
There are multiple ways to do that:
1. Script serves an HTTP API endpoint and liquidsoap tries to connect to it periodically to get the next track URL.
1. Liquidsoap serves an HTTP API endpoint and your script tries to connect to it periodically to push the next track URL.
1. Liquidsoap periodically runs a shell command, that will calculate time, get state, and print the track URL to stdout.

If you choose to run an HTTP server on the script side, then you may want to implement these functions:
1. `GET /progress` - to get the queue progress. Liquidsoap will get the queue and current position.
1. `PUT /progress` - to put the queue progress in the current track. Liquidsoap will update the queue and the current position.

It's possible not to implement the position at all and only serve something like `GET /next`.
The problem with this design is that you have to implement logic specific only for a single task on the liquidsoap side. The other method doesn't have it.

If you choose to run the HTTP server on the liquidsoap side, then the API endpoints may look like
1. `POST /queue/track` - to add the next track to the queue.
1. `GET /queue/progress` - to get the progress of the queue.

The problems are
1. Directly after restart the queue will be empty.
1. Multiple scripts will require the implementation of some form of queue locking.

At the moment of writing these interactions, I'm trying to understand if the same could be achieved by using a simple function that will save the state in json format to a file and load it on start. So here's another plan

1. On boot load state and playlists.
1. If the time slot has passed and the position is not at the beginning - skip track.
1. Put track to the queue and change the position.
1. `switch([({8h00m}, q1), ({16h00m}, q2)])`.
1. Save positions to file periodically.
1. Periodically put track if the queue is empty.

**Reply by gAlleb** [reply]
Thanks a lot, Vito! I'll ponder my next steps.

**Reply by gAlleb** [reply]
Hi Vito!

> PUT /progress - to put the queue progress in the current track. Liquidsoap will update the queue and the current position.  It's possible not to implement the position at all and only serve something like GET /next.
The problem with this design is that you have to implement `logic specific only for a single task on the liquidsoap side. The other method doesn't have it`

Didn't get this bit.

So, I've implemented this on LS side:

```ruby
radio = single("1.mp3")

api_url = "http://api:9999"
auth_api_url = string.base64.encode("login:pass")

# Made POST method api (instead of GET) cause it can have broad functionality if needed

def api_call(~timeout=2.0, url, payload) =

    full_url = "#{api_url}/#{url}"

    log("API #{url} - Sending request to '#{full_url}' with body: #{payload}")
    try
        response = http.post(full_url,
            headers=[
                ("Content-Type", "application/json"),
                ("Authorization", "Basic #{auth_api_url}")
            ],
            timeout=timeout,
            data=payload
        )

        log("API #{url} - Response (#{response.status_code}): #{response}")
        "#{response}"
    catch err do
        log("API #{url} - Error: #{error.kind(err)} - #{error.message(err)}")
        "false"
    end

end

# Ten tracks queue
ten_queue = request.queue()

def next_song() =
    response = api_call(
        "nextTrack", ""
    )
    if (response == "") or (response == "false") then
        null()
    else
	ten_queue.push(request.create(response))
    end
end

thread.when(predicate.activates({08h00m or 16h00m}), next_song)

# made it "true" on purpose 
radio = fallback(track_sensitive=true, [ten_queue, radio])

output(mksafe(radio))
```

And some js on server side:

```js
let playlistTracks = [];
let numberOfTracks;
let currentTrackIndex = getCurrentTrackIndex();

function importPlaylistFromFile(pathToFile) {
    try {
        const data = fs.readFileSync(pathToFile, 'utf8');
        playlistTracks = data.trim().split('\n');
        numberOfTracks = playlistTracks.length;
        return playlistTracks;
    } catch (err) {
        console.error('Error reading the playlist file:', err);
        return [];
    }
}

// Watch for changes in the playlist file
function watchPlaylistFile(filePath) {
    importPlaylistFromFile(filePath);
    const watcher = chokidar.watch(filePath);
    watcher.on('change', () => {
        console.log('Playlist has been modified.');
        importPlaylistFromFile(filePath);
        console.log('Playlist tracks:');
        console.log(playlistTracks);
        console.log('Number of tracks:');
        console.log(numberOfTracks);
        console.log('Current track:');
        console.log(getTrackPath());
        console.log('Current track index:');
        console.log(currentTrackIndex);
    });
}

// Path to the playlist file
const playlistUrl = 'test_playlist.m3u';

// Watch the playlist file for changes
watchPlaylistFile(playlistUrl);

// Here we can set the startDate - so it's a constant state which makes it easier if we we reload
function getCurrentTrackIndex() {
    const startDate = new Date('2024-07-18');
    const currentDate = new Date();
    const elapsedTime = currentDate - startDate;
    const dayInMS = 1000 * 60 * 60 * 24;
    return Math.floor(elapsedTime / dayInMS);
}

// Each day only 1 file path is available, next day it changes 

function getTrackPath() {
    trackIndex = currentTrackIndex % numberOfTracks;
    return playlistTracks[trackIndex];
}

app.post('/nextTrack', (req, res) => {
    const nextTrackPath = getTrackPath();
    res.send(nextTrackPath);
});

```

**Reply by vitoyucepi** [reply]
Maybe hide the complexity?
```ruby
ten_queue = request.queue()

# TODO: rename method
def next_song()
  # api.get should either return null() or data
  response = api.get.next()
  if response != null() then
    ten_queue.push(request.create(null.get(response)))
  end
end

thread.run(next_song, every=60.)
```

Maybe use a token instead, say `${well-known-prefix}-${random(20)}`?
```ruby
api_auth_token = environment.get("API_TOKEN")
headers=[("PRIVATE-TOKEN", api_auth_token)]
```

> Didn't get this bit.

Which part of it? The progress?

1. Periodically send the current track and its position to the server.
1. After restart, attempt to restore the track position by reporting that you have lost the progress.

**Reply by gAlleb** [reply]
> thread.run(next_song, every=60.)

But why? We need to play track 2 times a day. So push it to the queue 2 times a day. If we run it every minute - it will constantly push the track.


> Periodically send the current track and its position to the server.
    After restart, attempt to restore the track position by reporting that you have lost the progress.

I didn't get to understand how to do it. Could you elaborate?

And I've made a static start position (`const startDate = new Date('2024-07-18');`) so progress can't be lost. So is the position needed?

**Reply by vitoyucepi** [reply]
```ruby
def daily_playlist()
  def api(method, path, data=null())
    response =
      if method == "get" then
        http.get("#{url}/#{path}", headers=headers)
      elsif method == "put" then
        http.put("#{url}/#{path}", headers=headers, data=data)
      else
        null()
      end
    null.map(is_valid, response)
  end

  def api.get(path)
    api("get", path)
  end

  def api.put(path, data)
    api("put", path, data)
  end

  def api.get.next()
    response = api.get("next")
    # TODO: parse response
    response
  end

  def api.get.progress()
    response = api.get("progress")
    # TODO: parse response
    response
  end

  def api.put.progress(data)
    # TODO: reformat data
    response = api.put("progress", data)
    # TODO: parse response
    ignore(response)
  end

  queue = request.queue()

  def get_next()
    next = api.get.next()
    if next != null() then
      queue.push(request.create(null.get(next)))
    end
  end

  def restore_progress()
    progress = api.get.progress()
    if progress.track != null() then
      # TODO: check edge cases
      queue.push(request.create(null.get(progress.track)))
      queue.seek(progress.position)
    end
  end

  def put_progress()
    current = queue.current()
    elapsed = queue.elapsed()
    if current != null() then
      # TODO: extract current track
      api.put.progress({current = current, elapsed = elapsed})
    end
  end

  restore_progress()
  thread.run(get_next, every=60.)
  thread.run(put_progress, every=10.)
end
```

**Reply by gAlleb** [reply]
Thanks!

**Reply by gAlleb** [reply]
So after carefully looking at and doing some magic movements with ma mind - I'm struggling to see the whole picture. 

1. I presume that I should then rewrite the logic of my API side? Currently all it does - it serves everyday new `file path` according to schedule on one end point. And here we have 3 (next (get), progress (get, put)). 
- next (get) - get track, that is clear.
- progress (put) - Periodically send the current track and its position to the server.
- progress (get) - After restart, attempt to restore the track position by reporting that you have lost the progress. 
Last two - not so clear. Why should I send the current track and its position to the server if It's the server which sends the track to LS in the first place?

Or it means that I have to push `the whole playlist` to the queue with a number so then it could be processed (get the number, skip tracks before it)?

2. I still don't get `thread.run(get_next, every=60.)`. It would constantly push the same track during the day? OR I suppose to feed the queue in another way and add this logic to API? Should I add switches with `predicate`?

3.  Probably I'm missing something due to my lack of knowledge (or mere understanding or both)). E.g. `queue.seek(progress.position)` We seek within our queue source but doesn't `seek` seeks in seconds? Or we seek the current playlist in the queue with some kind of position number?

Actually I got a bit confused) Feel free to ignore this message as I do understand that it might cause frustration (or irritation of some sort)).

**Reply by vitoyucepi** [reply]
> Why should I send the current track and its position to the server if It's the server which sends the track to LS in the first place?

I intend to send feedback to the script that will store the progress.

> It would constantly push the same track during the day?

You shouldn't send tracks every time. If you receive the track from the server using `next`, then API will mark it as consumed.

> We seek within our queue source but doesn't seek seeks in seconds

Yeah, seek in seconds. I don't know what you are trying to implement overall, but I imagined a podcast app that will preserve the playback progress even after the device restarts.

**Reply by gAlleb** [reply]
> Yeah, seek in seconds. I don't know what you are trying to implement overall, but I imagined a podcast app that will preserve the playback progress even after the device restarts.

A-haa! Now I get what you mean. That is good. Gonna ponder. Thank you, Vito!

**Reply by gAlleb** [reply]
Am I wrong somewhere or what? Can't get anything out of `queue.current()` (2.3.x, 2.2.5)

```     
    current = queue.current()
    elapsed = queue.elapsed()
    log("#{current} #{elapsed}")
```

```
2024/07/24 19:03:40 [lang:3] <request(id=3)> 89.6
2024/07/24 19:03:50 [lang:3] <request(id=3)> 99.56
2024/07/24 19:04:00 [lang:3] <request(id=3)> 109.6
2024/07/24 19:04:10 [lang:3] <request(id=3)> 119.6
```

Or is supposed to return `request(id=3)`? I thought it should return filename.

**Reply by vitoyucepi** [reply]
> I thought it should return filename.

`request.metadata(current)`, `request.filename(current)` or `request.uri(current)`.

**Reply by gAlleb** [reply]
Thanks, Vito!
```
request.uri(current)

Error 5: this value has type
  _? (inferred at test.liq, line 100, char 14-29)
but it should be a subtype of
  request
```

**Reply by vitoyucepi** [reply]
Use `null.get`, the `queue.current()` returns nullable.

**Reply by gAlleb** [reply]
```
    current_uri = null.get(request.uri(current))

Error 5: this value has type
  _? (inferred at test_api.liq, line 103, char 14-29)
but it should be a subtype of
  request
```
If I use `null.get` on `queue.current()` I get
```
[runtime:1] Uncaught error error(kind="not_found",message="no default value for null.get",positions="at null.liq, line 13, char 16-77, at null.liq, line 13, char 16-77, at null.liq, line 13, char 5-77, at null.liq, line 12 char 2 - line 14 char 17, at test_api.liq, line 104, char 30-47, at thread.liq, line 10, char 11-14")!
```

**Reply by vitoyucepi** [reply]
> `null.get(request.uri(current))`
> `no default value for null.get`

You should first check that `current` is not `null`, and then get the uri.
You can't get the uri from `null` because it's not a request.

```ruby
if current != null() then
  current_uri = request.uri(null.get(current))
  # do something
end
```

I would probably use a `null.map` for that. The type of `current_uri` is `string?`, and in some cases the previous code snippet may be more desirable.

```ruby
current_uri = null.map(request.uri, current)
```

**Reply by gAlleb** [reply]
So I've managed to make it work with some alterations.

1. I've commented out `def daily_playlist()` because I don't know how then to pass `queue` to fallback.
2. I've changed `api` function to 2 separate functions because it didn't want to work)
3. I've parsed `progress` with `let json.parse ({current, elapsed} : {current: string, elapsed: float}) = progress` because it doesn't parse with `let json.parse parsed_json = progress` so we could have `parsed_json.current and parsed_json.elapsed`
4. Had to embrace  queue.seek(elapsed) in `thread.run` and add delay - ` thread.run(delay=0.5, {ignore(queue.seek(elapsed))})` because otherwise it refuses to seek.
5. Had to add delay and thread.run -> `thread.run(restore_progress, delay=0.5)`. Otherwise it won't play when LS is restarted. A random track from main playlist kick's in.
6. Also I've noticed that `elapsed` always starts at 0 even if we restart LS, get track and seek forward 120 sec.
7. Also I was thinking about sending null when track is finished
```ruby
    current = null.map(request.uri, queue.current())
    elapsed = queue.elapsed()
    
    if current != null() then
      log("#{current} #{elapsed}")
      # TODO: extract current track
      ignore(api_put.progress({current = current, elapsed = elapsed}))
    else
      ignore(api_put.progress({current = null_string(current), elapsed = null_float(elapsed)}))
    end
```
and adding delay to ` thread.run(put_progress, every=10., delay=10.)` in order not get empty progress when LS restarts during playback. Because if it restarts during such playback - 1st thing it will send null progress and then fetch it instead of a real one. Why to send null after track is finished? Here is the case:
- we retrieve 3 hour long track and play it, in 2.5 hours LS restarts and retrieves progress in order to restore it.
- 3 hour long track starts playing from 2.5 hour mark and stops within 30 minutes because it has reached the end. `elapsed` starts counting from `0` when we seek the queue. So when track has finished playing we will have `elapsed: 30 minutes` stored.
- In one more hour after track has finished playing LS restarts again (who knows why, just an edge case) and retrieves progress. But now we have only 30 minute progress data (because LS has been previously restarted and seeked through queue).
- And track starts to play 30 minutes into an audio file. And will play another 2.5 hours.

<hr>

```ruby
radio = playlist("music.m3u8")
api_url = "http://api"
#def daily_playlist()
 def api_put(path, data) =

    full_url = "#{api_url}/#{path}"

    log("API #{path} - Sending POST request to '#{full_url}' with body: #{data}")
    try
        response = http.put(full_url,
            headers=[
                ("Content-Type", "application/json")
            ],
            data=json.stringify(data)
        )

        log("API #{path} - Response (#{response.status_code}): #{response}")

	response
    catch err do
        log("API #{path} - Error: #{error.kind(err)} - #{error.message(err)}")
        "false"
    end

end

def api_get(~timeout=2.0, path) =

    full_url = "#{api_url}/#{path}"

    log("API #{path} - Sending GET request to '#{full_url}'")
    try
        response = http.get(full_url,
            headers=[
                ("Content-Type", "application/json")
            ],
            timeout=timeout
        )

        log("API #{path} - Response (#{response.status_code}): #{response}")
        "#{response}"
    catch err do
        log("API #{path} - Error: #{error.kind(err)} - #{error.message(err)}")
        "false"
    end

end

 
  def api_get.next()
    response = api_get("next")
    # TODO: parse response
    response
  end

  def api_get.progress()
    response = api_get("progress")
    # TODO: parse response
    response
  end

  def api_put.progress(data)
    # TODO: reformat data
    response = api_put("progress", data)
    # TODO: parse response
    ignore(response)
  end

  queue = request.queue()

  def get_next()
    next = api_get.next()
    if next != null() then
      queue.push(request.create(null.get(next)))
    end
  end

  def restore_progress()
    progress = api_get.progress()

    if progress != "" then
    log("#{progress}")
    let json.parse ({current, elapsed} : {current: string, elapsed: float}) = progress
    log("#{current}: #{elapsed}")
      # TODO: check edge cases
     if elapsed != null() then
      queue.push(request.create(null.get("#{current}")))
      thread.run(delay=0.5, {ignore(queue.seek(elapsed))})
     end
    end
  end

  def put_progress()

    current = null.map(request.uri, queue.current())
    elapsed = queue.elapsed()
    
    if current != null() then
      log("#{current} #{elapsed}")
      # TODO: extract current track
      ignore(api_put.progress({current = current, elapsed = elapsed}))
    end
  end

  thread.run(restore_progress, delay=0.5)
  thread.run(get_next, every=60.)
  thread.run(put_progress, every=10.)
#end

radio = fallback(track_sensitive=false, [queue,radio])
 
output(mksafe(radio))
```

simple js. When playlist is imported - becomes an array with 2 consumed flags:
```
{"path":"1.mp3","consumed":false,"consumed2":false}
```

```js
app.get('/next', checkDayChange, (req, res) => {

    const currentTrack = getTrackPath();
    
    // Check if the current track has been consumed

    if (!currentTrack.consumed && (new Date().getHours() >= 8 && new Date().getHours() < 9)) {

        // Mark the track as consumed
        currentTrack.consumed = true;
        // Return the path of the current track
        res.send(currentTrack.path);

    } else if (!currentTrack.consumed2 && (new Date().getHours() >= 16 && new Date().getHours() < 17)) {
        currentTrack.consumed2 = true;
        res.send(currentTrack.path);
    } else {

        return res.send("");

    };
});
let progress_data;
app.put('/progress', (req, res) => {
progress_data = req.body;
res.sendStatus(200);
});
app.get('/progress', (req, res) => {
res.send(progress_data);
});
```

I don't know why but when I use `put` endpoint - I get `100` response in LS. With `post` I get `200` as usual.

**Reply by gAlleb** [reply]
Regarding `elapsed starts counting from 0 when we seek the queue after progress is restored` and an edge case I came up with this:

```ruby
restored_elapsed = ref(0.0)

  def restore_progress()
    progress = api_get("progress")

    if progress != "" then
    
    let json.parse ({current, elapsed} : {current: string, elapsed: float}) = progress

     if elapsed != null() then

      restored_elapsed := elapsed

      queue.push(request.create(null.get("#{current}")))
      thread.run(delay=0.5, {ignore(queue.seek(elapsed))})
     end
    end
  end

  def put_progress()

    current = null.map(request.uri, queue.current())
    elapsed = queue.elapsed()
    
    if current != null() then
      log("#{current} #{elapsed}")
      # TODO: extract current track
      ignore(api_put.progress({current = current, elapsed = elapsed + restored_elapsed()}))
    end
  end

```

We have a `restored_elapsed = ref(0.0)` that gets populated from restored progress and then we put elapsed with addition of `restored_elapsed`.

Vito, what do you think about it? Maybe you got some insights on that or on little problems I had regarding `def daily_playlist()` that I had to comment out because I don't know how then to pass queue to fallback)

**Reply by vitoyucepi** [reply]
> Maybe you got some insights on that or on little problems I had regarding `def daily_playlist()` that I had to comment out because I don't know how then to pass queue to fallback

The last line of the function acts as an implicit return.

```ruby
def daily_playlist()
  queue = request.queue()

  # do something

  queue
end
```

You may want to return some other methods
```ruby
def daily_playlist()
  queue = request.queue()

  def some_function()

    # do something

  end

  # do something

  {queue = queue, f = some_function}
  # queue.{f = some_function}
end
```

**Reply by gAlleb** [reply]
Thank you, Vito!
